### ISSUE TITLE
Respect `rust-version` when selecting a version for `cargo install`

### ISSUE THREAD

**Issue by epage** [rust-lang/cargo#10903]
When running `cargo install foo`, if my rust version is too old, I'll get compilation errors.  If the failing crate sets `rust-version`, then I'll get  a nice error message about it but no path forward for fixing it.  #10891 adds a path forward for `cargo install` when the failure is in a dependency but not when its in the crate itself that is being installed.

Ideally, `cargo install` will select an older version (with a warning) that is compatible with my toolchain.  As a step back from the ideal, I'd get an error with corrective steps (ideally, telling me what version to use).

**Comment by epage** [comment]
This is blocked on `rust-version` being in the registry

**Comment by 8573** [comment]
While I generally very much support supporting older Rust versions, I worry about how this would (not silently but) quietly give users insecure versions of programs, in the absence of either integration with RustSec or a culture of yanking insecure versions (which I would hope would be blocked on having [a way to bypass yankedness](https://internals.rust-lang.org/t/cargo-allow-publishing-the-reason-for-yanking/16896/22) for e.g. `git bisect`).

**Comment by epage** [comment]
Our choices are
- Error and let the user figure it out (current behavior)
- Error and tell the user how to resolve it
  - This is included as an alternative in this issue
  - This puts it in the users hands for deciding what is safe for them
  - We are still leaving a road hump along the users path to accomplishing their goal, souring their opinion of the ecosystem and possibly being the part that broke their will to continue experimenting with the solution (which will include the common refrain "why is it erroring instead of just doing it")
- Auto-select a compatible version
  - This is included as the primary proposed solution to this issue
  - The user might be surprised at the version used and lack of features
  - The crate version *might* have vulnerabilities that affect the user

Maybe I'm missing some use cases but I personally have little sympathy for the security side of this.  If the error just leads them to selecting an old version anyways, did we really help them with security?  How do we know the newest version is any more secure than the older version?  This also isn't too different than if someone had `cargo install`ed things previously.

**Comment by cassaundra** [comment]
Thoughts I had from a discussion with Ed:

My inclination here is to have `cargo install` still error, but provide a more useful suggestion. I see a couple of problems with auto-selecting a compatible version:
- Most users are running the latest Rust stable/nightly installed via rustup (per [2021 survey](https://raw.githubusercontent.com/rust-lang/surveys/main/surveys/2021-annual-survey/2021-annual-survey-summary.pdf)), so an out-of-date compiler is more likely to signal to the user that they should upgrade their installation, rather than install an older version of a package.
- `cargo install` is pretty noisy already, so users might miss the warning.
  - The warning could be put at the end, but by that time the package has already been installed.
  - Even if made less noisy, the warning does not give the user much time to decide what to do.

**Comment by epage** [comment]
"programs you run" is a little different than "project dependencies" and can more likely be ahead.

I also understand the problem of `cargo install`s output.  In general, I wonder if we can do things to shrink our build and test output but that is outside of the scope of this.

Telling the user what version they can install would be a good first step for improving the error message.  As that is just UX, we can do that without putting it behind the `-Zmsrv-policy` feature flag.  We can always explore auto-selecting for cargo install in the future.

**Comment by epage** [comment]
FYI #12553 should make further MSRV work a bit easier

**Comment by charles-r-earp** [comment]
> Telling the user what version they can install would be a good first step for improving the error message. As that is just UX, we can do that without putting it behind the `-Zmsrv-policy` feature flag. We can always explore auto-selecting for cargo install in the future.

Is it possible to implement `cargo +nightly install -Z msrv-policy`?

Will this be the default when stable, or is it opt in, behind resolver 3 (#9930)?

**Comment by epage** [comment]
`cargo install` was left out of scope for the MSRV-aware Cargo work, besides the error improvement mentioned earlier, as there were additional design concerns to be worked out and we didn't want to smaller improvement on this.

**Comment by JustusFluegel** [comment]
I would really love if there is at least a flag, `--auto-downgrade` or something (which still prints out a warning, just not aborting) since i run cargo install <foo> in some places where we don't want to pin the version in ci but our rust builds failing on older versions (i generally run ci on nightly, beta, stable, and 2-3 fixed "semi-up-to-date" versions (for example 1.79 1.78 and 1.77 currently)) is a pretty bad experience. Adding a `--auto-downgrade` or so would make that use-case way simpler.

**Comment by johntyree** [comment]
Thinking that "most users" is the same as "users that would bother filling out our survey" is kind of silly, honestly.

**Comment by pksunkara** [comment]
@epage Are there any blockers with allowing an `--auto-downgrade` option? I agree that it's much needed in ci systems.

I would be willing to implement it if allowed.

**Comment by epage** [comment]
Its not a question of "any blockers?" but "is that the right design?".

We need to be careful when making a decision like this.  Our CLI is part of our stability guarantees and we are limited in how we can evolve them.  We also have to deal with the lesson I learned from clap: the larger an "API" surface, the harder it is to discover what features exist, the less value *all users* get out of *every* feature when adding a new one.

If we do more, the question is
- Opt-in or opt-out
- How is it specified

For example, `cargo add` is opt-out with `--ignore-rust-version`.  There is also a suggestion for `cargo add foo@latest` with an open question on whether that would ignore MSRV (#10741).

However, `cargo add` making a bad choice has less of an impact; just revert.  The impact of operations is on a spectrum, from `cargo check`, with little to no side effects and is fast, making it easy to recover from, to `cargo publish`, which is irreversible and can only be fixed up with `cargo yank`.  `cargo install` is somewhere in the middle.  The build times are slow enough that recovery is costly.  Currently, the output is noisy, losing warnings (though that would be fixed by #8889).  If its run in CI, users will never see a warning until they observe behavior that surprises them and then they have to debug their process until they find it.  End-users are also less likely to be actively running the old version and may rather error if they accidentally are (e.g. through `rust-toolchain.toml` files).  Use of older versions is more likely coming through specialized processes, so there is less benefit to smoothing out the workflow with defaulting to MSRV-compatible version selection with an opt-out.

The main benefit to defaulting to MSRV-compatible version selection with an opt-out is consistency

So if we allow MSRV-compatible version selection through an opt-in, the questions are
- How do we name it?  Currently, MSRV-aware resolver uses `incompatible-rust-versions = "<allow|fallback>"`
- Where do we put the opt-in?  We have CLI flag, config, maybe even `@<something>`
- How will it interact with `--ignore-rust-version`?
